pkg/compactor/planned_jobs_http.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/oklog/ulid"
+	"github.com/prometheus/prometheus/model/timestamp"
+
+	"github.com/grafana/mimir/pkg/storage/bucket"
+	"github.com/grafana/mimir/pkg/storage/tsdb"
+	"github.com/grafana/mimir/pkg/storage/tsdb/bucketindex"
+	"github.com/grafana/mimir/pkg/util"
+)
+
+//go:embed compactor_tenants.gohtml
+var tenantsPageHTML string
+var tenantsTemplate = template.Must(template.New("webpage").Parse(tenantsPageHTML))
+
+type tenantsPageContents struct {
+	Now     time.Time `json:"now"`
+	Tenants []string  `json:"tenants,omitempty"`
+}
+
+func (c *MultitenantCompactor) TenantsHandler(w http.ResponseWriter, req *http.Request) {
+	tenants, err := tsdb.ListUsers(req.Context(), c.bucketClient)
+	if err != nil {
+		util.WriteTextResponse(w, fmt.Sprintf("Can't read tenants: %s", err))
+		return
+	}
+
+	util.RenderHTTPResponse(w, tenantsPageContents{
+		Now:     time.Now(),
+		Tenants: tenants,
+	}, tenantsTemplate, req)
+}
+
+//go:embed planned_jobs.gohtml
+var plannerJobsHTML string
+var plannerJobsTemplateFuncs = template.FuncMap{"add": func(x, y int) int { return x + y }}
+var plannerJobsTemplate = template.Must(template.New("webpage").Funcs(plannerJobsTemplateFuncs).Parse(plannerJobsHTML))
+
+type plannerJobsContent struct {
+	Now                string `json:"now"`
+	BucketIndexUpdated string `json:"bucket_index_updated"`
+
+	Tenant      string                 `json:"tenant"`
+	PlannedJobs []plannedCompactionJob `json:"jobs"`
+
+	ShowBlocks     bool `json:"-"`
+	ShowCompactors bool `json:"-"`
+
+	SplitJobsCount int `json:"split_jobs_count"`
+	MergeJobsCount int `json:"merge_jobs_count"`
+
+	TenantSplitGroups int `json:"tenant_split_groups"`
+	TenantMergeShards int `json:"tenant_merge_shards"`
+	SplitGroups       int `json:"-"`
+	MergeShards       int `json:"-"`
+}
+
+type plannedCompactionJob struct {
+	Key       string      `json:"key"`
+	MinTime   string      `json:"min_time"`
+	MaxTime   string      `json:"max_time"`
+	Blocks    []ulid.ULID `json:"blocks"`
+	Compactor string      `json:"compactor,omitempty"`
+}
+
+func (c *MultitenantCompactor) PlannedJobsHandler(w http.ResponseWriter, req *http.Request) {
+	vars := mux.Vars(req)
+	tenantID := vars["tenant"]
+	if tenantID == "" {
+		util.WriteTextResponse(w, "Tenant ID can't be empty")
+		return
+	}
+
+	if err := req.ParseForm(); err != nil {
+		util.WriteTextResponse(w, fmt.Sprintf("Can't parse form: %s", err))
+		return
+	}
+
+	showBlocks := req.Form.Get("show_blocks") == "on"
+	showCompactors := req.Form.Get("show_compactors") == "on"
+	tenantSplitGroups := c.cfgProvider.CompactorSplitGroups(tenantID)
+
+	tenantMergeShards := c.cfgProvider.CompactorSplitAndMergeShards(tenantID)
+
+	mergeShards := tenantMergeShards
+	if sc := req.Form.Get("merge_shards"); sc != "" {
+		mergeShards, _ = strconv.Atoi(sc)
+		if mergeShards < 0 {
+			mergeShards = 0
+		}
+	}
+
+	splitGroups := tenantSplitGroups
+	if sc := req.Form.Get("split_groups"); sc != "" {
+		splitGroups, _ = strconv.Atoi(sc)
```

Comment:
Same comment as above regarding potential error swallowing.

Reply:
same reply, yes, we ignore errors here. if user puts garbage into the field, we use 0.